egs/wsj/s5/steps/nnet3/ctc/train_lstm.sh

```diff
@@ -238,12 +242,38 @@ if  [ $stage -le -6 ]; then
   echo "$0: creating CTC transition model"
 
   num_phones=$(cat $lang/phones.txt | grep -v '^#' | tail -n +2 | wc -l) || exit 1;
-  # important not to mak
-  $cmd $dir/log/init_trans_model.log \
-    ctc-init-transition-model --num-phones=$num_phones \
-       $alidir/tree \
+  num_cutoffs=$(echo $cutoffs|wc -w)
+  rm -f $dir/0.*.ctc_trans_mdl
+  $cmd JOB=1:$num_cutoffs $dir/log/init_trans_model.JOB.log \
+    cutoff_array=\(0 $cutoffs\) '&&' \
+    ctc-init-transition-model  --state-count-cutoff2plus=\${cutoff_array[JOB]} \
+       --num-phones=$num_phones $alidir/tree \
       "ark:gunzip -c $alidir/ali.*.gz | ali-to-phones $alidir/final.mdl ark:- ark:- |" \
-       $dir/0.ctc_trans_mdl || exit 1;
+       $dir/0.JOB.ctc_trans_mdl || exit 1;
+  for i in `seq 1 $num_cutoffs`;do
+    trans_model=$dir/0.${i}.ctc_trans_mdl
+    echo -n "$trans_model "
+    ctc-transition-model-info --print-args=false $dir/0.${i}.ctc_trans_mdl | grep num-history-states
+  done | python -c "
+import sys
+import operator
+
+target_history_states=$target_history_states
+model_num_states = []
+for line in sys.stdin:
+  parts = line.split()
+  assert(parts[1] == 'num-history-states:')
+  model_num_states.append([parts[0], abs(target_history_states - int(parts[2]))])
+
+print min(model_num_states, key = operator.itemgetter(1))[0]
+  " > $dir/selected_ctc_trans_model
+  ctc_trans_model=$(cat $dir/selected_ctc_trans_model)
+  ( cd $dir;
+  rm -f 0.ctc_trans_mdl;
+  ln -s `basename $ctc_trans_model` 0.ctc_trans_mdl;
+  )
+  echo "Desired num_history_states is $target_history_states, selected"
+  ctc-transition-model-info $ctc_trans_model | grep num-history-states
 fi
 
```

Comment:
Actually, Vijay, maybe hold off on this patch-- when I see this implemented I see how painful it can be.
Let me implement a line-search in C++ that will do this more easily.